server/src/main/java/org/elasticsearch/cluster/metadata/AliasOrIndex.java

```diff
@@ -134,8 +149,11 @@ public AliasMetaData getFirstAliasMetaData() {
             return referenceIndexMetaDatas.get(0).getAliases().get(aliasName);
         }
 
-        void addIndex(IndexMetaData indexMetaData) {
+        void addIndex(IndexMetaData indexMetaData, @Nullable Boolean isWriteIndex) {
```

Comment:
Instead of doing this like this maybe only do it at the end? add a method called "computeAndValidateWriteIndex" and let that one do two things:
- scan the all index metadata and extracts the writeIndex, if any
- validate we have no two indices which are marked with the index_write flags.

you can use org.apache.lucene.util.SetOnce to store the writeIndex reference. Then you can assert that once `computeAndValidateWriteIndex` is called you can't call addIndex anymore.